File: android/src/main/java/com/autoskeleton/AutoskeletonShimmerClock.kt
package com.autoskeleton

import java.util.UUID

// Task 4.4 (tasks.md Phase 4) / plan.md §3.6, ADR-8: Kotlin mirror of the TS
// `ShimmerClock` contract (`src/core/contracts.ts`), same structure as iOS's
// `AutoskeletonShimmerClock.swift`. ADR-8's whole point is "phase derived from an
// absolute origin, never from per-frame ticks" — every renderer instance reads the
// SAME `startedAt` epoch-ms origin and computes its own shader translation from
// `phaseAt`/`phaseOffsetMs`, so every shape and every list cell stays in phase with
// zero cross-instance coordination.

/** Normalized shimmer phase in [0, 1). Mirrors `ClockPhase` in `src/core/contracts.ts`. */
typealias AutoskeletonClockPhase = Double

class AutoskeletonShimmerClock(
    val id: String = UUID.randomUUID().toString(),
    periodMs: Double = 1500.0,
    /** `elapsedRealtime`, NOT `currentTimeMillis`.
     *
     *  The phase is `(now - startedAt) % periodMs`, so the clock only needs a
     *  monotonic origin — and wall-clock time is not one. An NTP correction or a
     *  manual clock change mid-animation shifts `now` without shifting
     *  `startedAt`, and every mounted overlay jumps to a new phase in the same
     *  frame. `elapsedRealtime` counts since boot, including deep sleep, and
     *  never steps.
     *
     *  iOS has never had this exposure: its running `CABasicAnimation` is based
     *  on `CACurrentMediaTime`, which is already monotonic. */
    private val now: () -> Double = { android.os.SystemClock.elapsedRealtime().toDouble() },
) {
    /** Android tier-1 always ticks via `Choreographer`/`postInvalidateOnAnimation`
     *  (brief §4 "Renderers > Default"), never per-frame JS. */
    val driver = "choreographer"
    val startedAt: Double = now()

    var periodMs: Double = periodMs
        private set
    var isPaused = false
        private set

    /** The clock's own reading of "now", in the SAME time base as [startedAt].
     *
     *  Exists so a renderer cannot pick its own source and silently drift: the
     *  draw pass used to call `System.currentTimeMillis()` while the origin came
     *  from the constructor's `now`, so overriding one without the other — which
     *  every test that injects a fake `now` does — produced a phase computed from
     *  two unrelated epochs. `phaseAt` is still public for callers that genuinely
     *  have their own timestamp. */
    fun nowMs(): Double = now()

    /** Pure function of time — used by the renderer on every draw and directly by
     *  tests. */
    fun phaseAt(timestampMs: Double): AutoskeletonClockPhase {
        val elapsed = timestampMs - startedAt
        val wrapped = normalizeRemainder(elapsed, periodMs)
        return wrapped / periodMs
    }

    /** Negative offset a joining renderer could use to start already in phase.
     *  Kept for parity with the TS/iOS contracts even though this renderer's own
     *  draw loop uses [phaseAt] directly rather than a native animation API with a
     *  `beginTime`/`animation-delay` concept. */
    fun phaseOffsetMs(nowMs: Double): Double {
        val elapsed = normalizeRemainder(nowMs - startedAt, periodMs)
        return -elapsed
    }

    fun setPeriod(ms: Double) {
        periodMs = ms
    }

    fun pause() {
        isPaused = true
    }

    fun resume() {
        isPaused = false
    }

    private fun normalizeRemainder(value: Double, modulus: Double): Double {
        val wrapped = value.rem(modulus)
        return if (wrapped < 0) wrapped + modulus else wrapped
    }
}

File: ios/AutoskeletonShimmerClock.swift
import Foundation
import QuartzCore

// Task 3.2 (tasks.md Phase 3) / plan.md §3.6, ADR-8: Swift mirror of the TS
// `ShimmerClock` contract (`src/core/contracts.ts`). ADR-8's whole point is "phase
// derived from an absolute origin, never from per-frame ticks" — every renderer
// instance reads the SAME `startedAt` epoch-ms origin and configures its own
// CoreAnimation `beginTime` from `phaseOffsetMs`, so every shape and every list cell
// stays in phase with zero cross-instance coordination and zero JS/native per-frame
// work on the production path.

/// Normalized shimmer phase in [0, 1). Mirrors `ClockPhase` in `src/core/contracts.ts`.
typealias AutoskeletonClockPhase = Double

protocol AutoskeletonClockTicking {
    /// `preferredFrameRateRange` for 120 Hz ProMotion (brief §4 "Shared shimmer
    /// clock"). NOTE: this does NOT drive the production shimmer animation itself —
    /// see `AutoskeletonShimmerClock`'s doc comment for why that must stay a pure
    /// `CABasicAnimation` instead. This display link exists only to (a) declare
    /// frame-rate intent for the display and (b) drive `subscribe()` callbacks,
    /// which the TS contract itself scopes to "DEV/DEBUG AND TESTS ONLY".
    func start(target: Any, selector: Selector)
    func stop()
}

/// Production `CADisplayLink`-backed ticker.
final class AutoskeletonDisplayLinkTicking: AutoskeletonClockTicking {
    private var displayLink: CADisplayLink?

    func start(target: Any, selector: Selector) {
        guard let target = target as? NSObject else { return }
        let link = CADisplayLink(target: target, selector: selector)
        link.preferredFrameRateRange = CAFrameRateRange(minimum: 60, maximum: 120, preferred: 120)
        link.add(to: .main, forMode: .common)
        displayLink = link
    }

    func stop() {
        displayLink?.invalidate()
        displayLink = nil
    }
}

/// One shared origin every `AutoskeletonRendererTier1` instance reads (ADR-8).
/// `driver` is always `"core-animation"` on iOS tier-1 — the mask/gradient shimmer
/// is a `CABasicAnimation` configured once at mount with a `beginTime` derived from
/// `phaseOffsetMs(now:)`, so it keeps animating even if the main run loop (and this
/// clock's own optional `CADisplayLink`) were starved. That decoupling is exactly
/// what makes NFR-2 (shimmer survives a blocked JS thread) true by construction
/// rather than by luck.
final class AutoskeletonShimmerClock: NSObject {
    let id: String
    let driver = "core-animation"
    private(set) var periodMs: Double
    private(set) var startedAt: Double
    private(set) var isPaused = false

    private let ticking: AutoskeletonClockTicking
    private let now: () -> Double
    private var subscribers: [UUID: (AutoskeletonClockPhase, Double) -> Void] = [:]
    private var isTicking = false

    init(
        id: String = UUID().uuidString,
        periodMs: Double = 1500,
        ticking: AutoskeletonClockTicking = AutoskeletonDisplayLinkTicking(),
        now: @escaping () -> Double = { Date().timeIntervalSince1970 * 1000 }
    ) {
        self.id = id
        self.periodMs = periodMs
        self.ticking = ticking
        self.now = now
        startedAt = now()
    }

    /// Pure function of time — used by renderers at mount and directly by tests.
    func phaseAt(_ timestampMs: Double) -> AutoskeletonClockPhase {
        let elapsed = timestampMs - startedAt
        let wrapped = elapsed.truncatingRemainder(dividingBy: periodMs)
        return (wrapped < 0 ? wrapped + periodMs : wrapped) / periodMs
    }

    /// Negative `beginTime` offset (in seconds, matching `CAMediaTiming`'s unit) a
    /// joining renderer uses so its `CABasicAnimation` starts already in phase with
    /// every other instance, instead of restarting the sweep from zero.
    func phaseOffsetMs(now nowMs: Double) -> Double {
        let elapsed = (nowMs - startedAt).truncatingRemainder(dividingBy: periodMs)
        return -(elapsed < 0 ? elapsed + periodMs : elapsed)
    }

    func setPeriod(_ ms: Double) {
        periodMs = ms
    }

    func pause() {
        isPaused = true
    }

    func resume() {
        isPaused = false
    }

    /// DEV/DEBUG AND TESTS ONLY per the TS contract — production rendering never
    /// ticks through this. Starts the display-link ticker lazily on first
    /// subscriber, stops it once the last one unsubscribes.
    func subscribe(_ listener: @escaping (AutoskeletonClockPhase, Double) -> Void) -> () -> Void {
        let token = UUID()
        subscribers[token] = listener
        if !isTicking {
            isTicking = true
            ticking.start(target: self, selector: #selector(handleTick))
        }
        return { [weak self] in
            self?.subscribers[token] = nil
            if self?.subscribers.isEmpty == true {
                self?.isTicking = false
                self?.ticking.stop()
            }
        }
    }

    @objc private func handleTick() {
        guard !isPaused else { return }
        let t = now()
        let phase = phaseAt(t)
        for listener in subscribers.values {
            listener(phase, t)
        }
    }
}
